Feature/dev 1837 and 1838

cookersjs opened the pull request:
Addresses [DEV-1837](https://reactome.atlassian.net/browse/DEV-1837) and [DEV-1838](https://reactome.atlassian.net/browse/DEV-1838).

This is a fairly large PR that introduces QA tests that flag various issues that were noticed during the CoV-2 annotations by Curation. These will be added to weekly QA.

New/Modified classes in org.gk.qualityCheck:

**NonHumanEventsNotManuallyInferredCheck** -- (DEV-1837) 

**HumanReactionsWithNonHumanPhysicalEntitiesWithoutDiseaseCheck** -- (DEV-1838-1) 

**NonHumanReactionsWithHumanPhysicalEntitiesCheck** -- (DEV-1838-2) 

**HumanReactionsWithoutDiseaseAndHaveNonHumanPhysicalEntities** -- (DEV-1838-3) 

**HumanReactionsWithNonHumanComplexesWithHumanComponentsCheck** -- (DEV-1838-4) 

**QACheckUtilities** -- Existing class that was modified. Common methods used by the above QA checks were added here. The especially important code to check here is the methods that **find all PEs in an event** (inputs, outputs, catalysts, regulators) and **skiplist checks** (recursive method for finding all parent Pathway DbIDs for a RlE and seeing if any of those DbIds are present in skiplist).

SolomonShorser-OICR commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+        return physicalEntities;
+    }
+
+    // Returns true if 'inferredFrom' referral is null.
```
Comment should be formatted as Javadoc. Also for `isHumanDatabaseObject`, `hasSpeciesAttribute`, `getInstanceAttributeName`; `setSkipList` & `setHumanSpeciesInst` are missing Javadoc.

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithNonHumanComplexesWithHumanComponentsCheck.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+
+    // Unused, but required, AbstractQualityCheck methods.
+    @Override
+    public void check() {
+    }
+
+    @Override
+    public void check(GKInstance instance) {
+    }
+
+    @Override
+    public void check(List<GKInstance> instances) {
+    }
+
+    @Override
+    public void check(GKSchemaClass cls) {
+    }
+
+    @Override
+    public void checkProject(GKInstance event) {
+    }
+
+    @Override
+    protected InstanceListPane getDisplayedList() { return null; }
```
It's a little strange to extend a class but not use the required inherited method.  Does this class have to extend from "AbstractQualityCheck"?

cookersjs replied:
Yes it was strange. It was a (not unprecedented) attempt to keep the structure of the QA checks in line with they were built for this project. This code is being moved to release-qa now though, so there won't be any need for these unused methods.

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithNonHumanPhysicalEntitiesWithoutDiseaseCheck.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.*;
+
+/**
+ * Flags all human Reactions that contain Non-only-Human PhysicalEntities that do not have a populated disease attribute.
```
I was a bit unsure of the term "Non-only-Human".  Does it mean Physical Entities that have at least one constituent part that is non-human?

cookersjs replied:
It's meant to convey either non-human species or non-populated species, I believe. I'll make that clearer (and less hypheny)

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithNonHumanPhysicalEntitiesWithoutDiseaseCheck.java:
```diff
@@ -0,0 +1,100 @@
+        QAReport report = new QAReport();
+        MySQLAdaptor dba = (MySQLAdaptor) dataSource;
+        QACheckUtilities.setHumanSpeciesInst(dba);
+        QACheckUtilities.setSkipList(Files.readAllLines(Paths.get("QA_SkipList/Manually_Curated_NonHuman_Pathways.txt")));
```
Could this be broken into multiple lines for readability 

```
QACheckUtilities.setSkipList(
    Files.readAllLines(
        Paths.get("QA_SkipList/Manually_Curated_NonHuman_Pathways.txt")
    )
);
```

or maybe create and call a simple helper method named something like "getNonHumanPathwaySkipList()"

`QACheckUtilities.setSkipList(getNonHumanPathwaySkipList());`

SolomonShorser-OICR replied:
Personally, I would avoid the multi-line version. It _looks_ like multiple expressions separated with `{` and `}` (because of the way it's indented), but it really is evaluated as one expression. Also, breaking up code like that makes me twitchy. Better to just have a few variable assignments. But this is a personal style thing, so... :shrug:

cookersjs replied:
I will implement `QACheckUtilities.setSkipList(getNonHumanPathwaySkipList());`

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithNonHumanPhysicalEntitiesWithoutDiseaseCheck.java:
```diff
@@ -0,0 +1,100 @@
+            for (GKInstance reactionPE : QACheckUtilities.findAllPhysicalEntitiesInReaction(reaction)) {
+                // Valid PhysicalEntities include those that have a non-human species OR have a human species AND have a relatedSpecies,
+                // and that do not have a populated disease attribute.
+                if ((QACheckUtilities.hasNonHumanSpecies(reactionPE) || hasHumanSpeciesWithRelatedSpecies(reactionPE))
+                        && reactionPE.getAttributeValue(ReactomeJavaConstants.disease) == null) {
```
Since the && has higher precedence, the last two conditions would get grouped to be the right-hand-side result of the ||.  Could the last two conditions be put in parentheses just to make it clear they are one unit?

SolomonShorser-OICR replied:
The first two conditions that are `||`d are in parentheses.

cookersjs replied:
Yea, its `(condition1 || condition2) && condition 3`

jweiser replied:
Sorry! I missed the parens around the `||`.  Don't listen to me regarding this :).

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithoutDiseaseAndHaveNonHumanPhysicalEntities.java:
```diff
@@ -0,0 +1,102 @@
+    public QAReport checkInCommand() throws Exception {
+        QAReport report = new QAReport();
+        MySQLAdaptor dba = (MySQLAdaptor) dataSource;
+        QACheckUtilities.setSkipList(Arrays.asList("168249")); // Innate Immunity Pathway
```
Could the db id for Innate Immunity Pathway be put into a named constant rather than a comment so it is part of the code?  I worry about the comment being de-synced from the value.

jweiser commented on src/org/gk/qualityCheck/HumanReactionsWithoutDiseaseAndHaveNonHumanPhysicalEntities.java:
```diff
@@ -0,0 +1,102 @@
+            // isHumanDatabaseObject checks that the species attribute only contains a Homo sapiens species instance. Multi-species RlEs are excluded.
+            if (!QACheckUtilities.memberSkipListPathway(reaction)
+                    && QACheckUtilities.isHumanDatabaseObject(reaction)
+                    && reaction.getAttributeValue(ReactomeJavaConstants.disease) == null) {
```
I notice this check has been used a couple of times.  Maybe it could get a named method in QACheckUtilities:

```
public boolean hasDiseaseTag(GKInstance instance) {
    return instance.getAttributeValue(ReactomeJavaConstants.disease) != null;
}
```
and then just add a "!" to check for the absence of disease.  

For me, having a name for the what the condition is checking makes it clearer at a glance.

cookersjs replied:
Yea I did consider that as well at the time, I'll implement that

jweiser commented on src/org/gk/qualityCheck/NonHumanEventsNotManuallyInferredCheck.java:
```diff
@@ -0,0 +1,80 @@
+        QAReport report = new QAReport();
+        MySQLAdaptor dba = (MySQLAdaptor) dataSource;
+        QACheckUtilities.setHumanSpeciesInst(dba);
+        QACheckUtilities.setSkipList(Files.readAllLines(Paths.get("QA_SkipList/Manually_Curated_NonHuman_Pathways.txt")));
```
Same as https://github.com/reactome/CuratorTool/pull/29/files#r429038750

jweiser commented on src/org/gk/qualityCheck/NonHumanReactionsWithHumanPhysicalEntitiesCheck.java:
```diff
@@ -0,0 +1,99 @@
+        QAReport report = new QAReport();
+        MySQLAdaptor dba = (MySQLAdaptor) dataSource;
+        QACheckUtilities.setHumanSpeciesInst(dba);
+        QACheckUtilities.setSkipList(Files.readAllLines(Paths.get("QA_SkipList/Manually_Curated_NonHuman_Pathways.txt")));
```
Same as https://github.com/reactome/CuratorTool/pull/29/files#r429038750

cookersjs commented:
I've addressed the PR comments that have been made so far. This new code is being moved to https://github.com/reactome/release-qa now, but feel free to comment here or in the new PR (once I open it) and I will address the comments. Core structure will remain (QA checks that rely heavily on a Utilities class), though the boilerplate surrounding it will obviously change.

jweiser commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+    public static Set<GKInstance> findAllCatalystPEs(GKInstance reaction) throws Exception {
+        Set<GKInstance> catalystPEs = new HashSet<>();
+        List<GKInstance> catalysts = reaction.getAttributeValuesList(ReactomeJavaConstants.catalystActivity);
+        for (GKInstance catalyst : catalysts) {
+            for (String attribute : Arrays.asList(ReactomeJavaConstants.activeUnit, ReactomeJavaConstants.physicalEntity)) {
+                for (GKInstance attributePE : (Collection<GKInstance>) catalyst.getAttributeValuesList(attribute)) {
```
Completely a style choice, but an alternative to 3 nested for loops could be to delegate the inner two loops to call a method that takes a catalyst instance and returns its physical entities:

`private List<GKInstance> getCatalystPhysicalEntities(GKInstance catalyst)`

cookersjs replied:
It's a good suggestion, but I think having a `getCatalystPEs` within a method already called `findAllCatalystPEs` makes the overall Utilities class a little more muddled. I'll just add a comment that says that a catalysts PEs come from its ActiveUnit and PhysicalEntity slots

jweiser commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+    public static Set<GKInstance> findAllRegulationPEs(GKInstance reaction) throws Exception {
+        Set<GKInstance> regulationPEs = new HashSet<>();
+        List<GKInstance> regulations = reaction.getAttributeValuesList(ReactomeJavaConstants.regulatedBy);
+        for (GKInstance regulation : regulations) {
+            for (String attribute : Arrays.asList(ReactomeJavaConstants.activeUnit, ReactomeJavaConstants.regulator)) {
+                for (GKInstance attributePE : (Collection<GKInstance>) regulation.getAttributeValuesList(attribute)) {
+                    // QA checks calling these methods are concerned with PhysicalEntities that have a species attribute.
```
Could use a helper method similar to the triple for loop in catalyst physical entities

cookersjs replied:
Same thing as I commented above for the similar catalyst situation

jweiser commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+     */
+    public static Set<GKInstance> findAllConstituentPEs(GKInstance multiPEInstance) throws Exception {
+        Set<GKInstance> physicalEntities = new HashSet<>();
+        if (multiPEInstance.getSchemClass().isa(ReactomeJavaConstants.Complex)) {
+            // Recursively search Complex components for all PhysicalEntities.
+            Collection<GKInstance> complexComponents = multiPEInstance.getAttributeValuesList(ReactomeJavaConstants.hasComponent);
+            for (GKInstance complexComponent : complexComponents) {
+                physicalEntities.addAll(findAllPhysicalEntities(complexComponent));
+            }
+        } else if (multiPEInstance.getSchemClass().isa(ReactomeJavaConstants.Polymer)) {
+            // Recursively search Polymer repeatedUnits for all PhysicalEntities.
+            Collection<GKInstance> polymerUnits = multiPEInstance.getAttributeValuesList(ReactomeJavaConstants.repeatedUnit);
+            for (GKInstance polymerUnit : polymerUnits) {
+                physicalEntities.addAll(findAllPhysicalEntities(polymerUnit));
+            }
+        } else if (multiPEInstance.getSchemClass().isa(ReactomeJavaConstants.EntitySet)) {
+            // Recursively search members and candidates (if EntitySet is a CandidateSet) for all PhysicalEntities.
+            physicalEntities.addAll(findEntitySetPhysicalEntities(multiPEInstance));
+        }
+        return physicalEntities;
```
The logic for finding physical entities in a complex and polymer is almost identical except for the attribute value of "hasComponent" and "repeatedUnit".  You could consolidate it into one method that takes the attribute to use to get the constituent as a parameter.  It would also simplify the body of this method so it's clear it's only switching based on whether the instance is a complex, polymer, or set.

cookersjs replied:
I joined them in a helper method. I considered using a ternary in the method:

`return complexOrPolymerInst.getSchemClass().isa("Complex") ? complexOrPolymerInst.getAttributeValuesList("hasComponent") : complexOrPolymerInst.getAttributeValuesList("repeatedUnit")`

But I decided to use the less concise if/elses because the ternary made Polymers more of an afterthought to the method

jweiser commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+    public static Set<GKInstance> findEntitySetPhysicalEntities(GKInstance entitySet) throws Exception {
+        Set<GKInstance> physicalEntities = new HashSet<>();
+        // Attribute values to be recursively searched for PhysicalEntities include 'hasMember' and 'hasCandidate'.
+        Set<String> entitySetAttributes = new HashSet<>(Arrays.asList(ReactomeJavaConstants.hasMember, ReactomeJavaConstants.hasCandidate));
+        // If the EntitySet is a DefinedSet, 'hasCandidate' attribute is not searched.
+        if (entitySet.getSchemClass().isa(ReactomeJavaConstants.DefinedSet)) {
+            entitySetAttributes.remove(ReactomeJavaConstants.hasCandidate);
+        }
```
It may be better to add candidate as an attribute to check if the entity set is a candidate set rather than remove it if the entity set is a defined set.  That way the attribute isn't initialized in the starting collection and then removed but only added when needed.

cookersjs commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+    // populated. If it isn't, returns null.
+    public static String getInstanceAttributeName(GKInstance instance, String attribute) throws Exception {
+        GKInstance attributeInstance = (GKInstance) instance.getAttributeValue(attribute);
+        return attributeInstance != null ? attributeInstance.getDisplayName() : null;
```
Funnily enough, this particular example is actually a case where the returned value is going to appear in the output report, and I wanted it to have 'null' so that it was visible. I'll keep this (or maybe returning 'N/A' makes more sense?) in this specific case, but I'll look through the code to make sure other cases like this return empty values

cookersjs replied:
I am going to change this method's name so that it ends with "ForOutputReport"

jweiser replied:
Yeah, I still think it makes sense to have a String object returned (empty or "N/A").  If you need it as a marker, a String describing what it is marking is preferrable to a `null` value in code.

Avoiding returning null values has become popular enough that there is actually a pattern to avoid it: https://en.wikipedia.org/wiki/Null_object_pattern#Java

cookersjs commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+     * @return Set<String> -- Once TopLevelPathway has been found, returns all DbIds, inclusive, between TopLevelPathway and original Event.
+     * @throws Exception -- Thrown by MySQLAdaptor.
+     */
+    private static Set<String> findEventHierarchyDbIds(GKInstance event, Set<String> dbIds) throws Exception {
+        dbIds.add(event.getDBID().toString());
+        Collection<GKInstance> hasEventReferrals = event.getReferers(ReactomeJavaConstants.hasEvent);
+        if (hasEventReferrals != null) {
+            for (GKInstance hasEventReferral : hasEventReferrals) {
+                dbIds.addAll(findEventHierarchyDbIds(hasEventReferral, dbIds));
+            }
+        }
+        return dbIds;
+    }
```
That works quite nicely

cookersjs commented on src/org/gk/qualityCheck/QACheckUtilities.java:
```diff
@@ -26,4 +33,297 @@ public static GKInstance getLatestCuratorIEFromInstance(GKInstance instance) thr
+     */
+    public static boolean memberSkipListPathway(GKInstance event) throws Exception {
+
+        if (skiplistDbIDs != null) {
+            // Finds all parent Event DbIds.
+            Set<String> hierarchyDbIds = findEventHierarchyDbIds(event, new HashSet<>());
+            // Check if any returned Event DbIds (including original Events) are in skiplist.
+            for (String skiplistDbId : skiplistDbIDs) {
+                if (hierarchyDbIds.contains(skiplistDbId)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
```
Oh nice, I was wondering if there was an easier way of doing that

SolomonShorser-OICR replied:
Is a NULL-check needed for `hierarchyDbIds`? If `findEventHierarchyDbIds` returns NULL, I think both methods will thrown an NPE.

jweiser replied:
@SolomonShorser-OICR The return value of `findEventHierarchyDbIds` is guaranteed to be an empty set if nothing is found, so a NPE shouldn't be a concern.
